clang/lib/APINotes/APINotesReader.cpp

```diff
@@ -373,6 +373,13 @@ void ReadFunctionInfo(const uint8_t *&Data, FunctionInfo &Info) {
       endian::readNext<uint16_t, llvm::endianness::little>(Data);
   Info.ResultType = std::string(Data, Data + ResultTypeLen);
   Data += ResultTypeLen;
+
+  unsigned SwiftReturnOwnershipLength =
+      endian::readNext<uint16_t, llvm::endianness::little>(Data);
+  Info.SwiftReturnOwnership = std::string(reinterpret_cast<const char *>(Data),
+                                          reinterpret_cast<const char *>(Data) +
+                                              SwiftReturnOwnershipLength);
```

Comment:
What do you think of making this stronger and using a `StringSwitch` to encode this as an enumerated type? That would compress the encoded serialisation, ensure that we catch an errant representation allowing better diagnostics, and just make this generally less confusing to see with the string representation.

Reply:
@compnerd my approach approach for `SwiftReturnOwnership` was to do something similar to `ResultType`.  Another reason for choosing `std:string` instead of `enum` was that `SwiftImportAs` is also implemented as `std:string`. `SWIFT_RETURNS_(UN)_RETAINED` annotation is closely related to `SWIFT_SHARED_REFERENCE` annotation. So I wanted to keep APINotes implementation of both of these similar. 

But I do agree that `enum` is a better approach for compression and diagnostics. But I also feel we'd not se much benefit unless we change other things like `SwiftImportAs` also to use `enums` for implementation. Do you think we should do such refactoring for all the annotations together in a separate follow-up patch?

Reply:
`ReturnType` is not the same category - the return type of a function can be any UDT. We have an explicit set of permissible ownership models.

Handling both `SWIFT_SHARED_REFERENCE` and this together does make sense.

Reply:
Makes sense @compnerd . I will start working on a follow up patch to convert other cases whic have a fixed/limited set of cases like SWIFT_SHARED_REFERENCE, etc to be implemented as Enums instead of `std:string`.